python/pyspark/sql/functions.py

```diff
@@ -1824,6 +1824,12 @@ class UserDefinedFunction(object):
     .. versionadded:: 1.3
     """
     def __init__(self, func, returnType, name=None):
+        if not callable(func):
+            raise TypeError(
+                "func should be a callable object "
```

Comment:
I think the error message could be more clear: "Not a function or callable (__call__ is not defined): {}"

I don't think it's necessary to use the name of the argument, but it isn't a problem if you think it is better.